### PR TITLE
Allow registering and executing WebAssembly functions

### LINKED ISSUE #17: Implement a WebAssembly runner function
A user-defined function capable of running WebAssembly code should be defined (and registered with `sqlite3_create_function`).

Its final specification should be subject to discussion, but the proposed format could look like that (pseudo-code):
```
fn run_wasm(wasm_module: blob, function_name: string, args...)
```

The function would accept a blob with compiled WebAssembly as its first argument, the Wasm function name to look up and execute as the second argument, and forward the rest of the arguments as is to this Wasm function to be executed.

This function can later be a foundation for allowing users to create and register arbitrary Wasm functions right from the CQL interface (esp. with some syntactic sugar to make it pretty: https://github.com/libsql/libsql/issues/18).

An important design decision is to pick a WebAssembly runtime for executing the modules. A preliminary list to consider:
 - wasmtime (https://github.com/bytecodealliance/wasmtime): feature-complete, I have some experience with it
 - wasmer (https://github.com/wasmerio/wasmer) reportedly faster
 - a generic interface which allows picking one, the other, or any other runtime

Right now I'm biased in favor of wasmtime, but since I saw @losfair's activity in this repository (welcome!!!) I can't resist asking - would you recommend wasmer instead? Any recommmendations for/against it?

### LINKED ISSUE #18: Implement CREATE FUNCTION for dynamic user-defined function creation
In order to provide best developer experience, it should be possible to create and register WebAssembly-based user-defined functions right from the CQL interface, without having to recompile libSQL or register it at the higher code layer, e.g. from a driver.

For that purpose, there already exists a `CREATE FUNCTION` statement and we should consider adding it to libSQL, with WebAssembly as the supported language. In order to be more or less human-readable, this statement could accept function body in WebAssembly text format (.wat), and then compile it locally, or just in the blob form, in case somebody prefers to avoid the recompilation.

Functions created by `CREATE FUNCTION` could be stored in a system table, 100% cached in memory, where they can be easily inspected and verified.

Then, the existing code paths for looking up SQL functions can be extended to also peek into the table mentioned above.

### PR DESCRIPTION
**Pull request by psarna**
This series implements a mechanism for registering and running Wasm functions. The current runtime of choice is wasmtime and its libwasmtime.so library with C bindings (but a switch to Rust should be considered, because that's the native language of wasmtime and the only interface which offers all of its features).

It operates on a very crude ABI (ref:#16), where ints and doubles are passed to/from WebAssembly as is,
and for strings/blobs/null it passes a pointer to a structure:
 - string: `[1 byte for type specification][data]`
 - blob: `[1 byte for type specification][4 bytes of size][data]`
 - null: `[1 byte for type specification]`

The way it's implemented now is twofold:
1. There's an internal `run_wasm` function, capable of running WebAssembly and translating the parameter types from and to the Wasm module
2. A dynamic lookup table, currently a regular SQL table: `CREATE TABLE libsql_wasm_func_table(name text PRIMARY KEY, body text)`. The table can be initialized from C code by calling `libsql_try_initialize_wasm_func_table()` or from shell by using a `.init_wasm_func_table` command.

After creating and filling the new meta-table, when a function call is used in a statement, e.g. `SELECT id, fib(id) FROM t`, and function `fib` is neither built-in nor user-defined, it will be looked up in the table. If found, its body will be assumed to hold valid WebAssembly code, compiled and run.

In order to enable WebAssembly integration, run configure with `./configure --enable-wasm-runtime` parameter.

A few examples WebAssembly-based user-defined functions coded in Rust can be found here: https://github.com/psarna/libsql_bindgen

Here's an inline demo for testing purposes, with a WebAssembly fibonacci sequence already compiled from Rust and copied in-place:
```sql
.init_wasm_func_table

CREATE FUNCTION fib LANGUAGE wasm AS '
(module 
 (type (;0;) (func (param i64) (result i64))) 
 (func $fib (type 0) (param i64) (result i64) 
 (local i64) 
 i64.const 0 
 local.set 1 
 block ;; label = @1 
 local.get 0 
 i64.const 2 
 i64.lt_u 
 br_if 0 (;@1;) 
 i64.const 0 
 local.set 1 
 loop ;; label = @2 
 local.get 0 
 i64.const -1 
 i64.add 
 call $fib 
 local.get 1 
 i64.add 
 local.set 1 
 local.get 0 
 i64.const -2 
 i64.add 
 local.tee 0 
 i64.const 1 
 i64.gt_u 
 br_if 0 (;@2;) 
 end 
 end 
 local.get 0 
 local.get 1 
 i64.add) 
 (memory (;0;) 16) 
 (global $__stack_pointer (mut i32) (i32.const 1048576)) 
 (global (;1;) i32 (i32.const 1048576)) 
 (global (;2;) i32 (i32.const 1048576)) 
 (export "memory" (memory 0)) 
 (export "fib" (func $fib)))
';

CREATE TABLE IF NOT EXISTS example(id int PRIMARY KEY);
INSERT OR REPLACE INTO example(id) VALUES (7);
INSERT OR REPLACE INTO example(id) VALUES (8);
INSERT OR REPLACE INTO example(id) VALUES (9);
SELECT id, fib(id) FROM example;
```

This series also comes with syntactic sugar for registering and deregistering Wasm functions dynamically via SQL: `CREATE FUNCTION` and `DROP FUNCTION`: Fixes #18

Fixes #17

### CODE REVIEW

**Comment by psarna** [comment]
This is only a draft for multitude of reasons, the most important ones being:
 - [ ] lack of automated tests
 - [ ] currently, invoking Wasm-based user-defined functions causes an explicit memory leak during lookup - these functions need to be tracked and cached (also to avoid Wasm recompilation)

**Comment by psarna** [comment]
@juntao I actually looked it up earlier today, we're definitely interested in giving it a go! And, eventually, make the implementation runtime-agnostic by relying on Wasm C API (https://github.com/WebAssembly/wasm-c-api) that @losfair mentioned in another issue.

I remember from my morning research that the C dynamic library from WasmEdge release page was ~50MB, which is quite heavy compared to libwasmtime's 17 - are you aware of any thinner versions of it?

**Comment by juntao** [comment]
Yes. I believe WasmEdge supports the standard C API -- I will confirm.

The WasmEdge dynamic library really should not be that big. The distribution binary of WasmEdge is only 8MB. I think the large version contains LLVM so that it can do AOT compilation w/o external dependency. Let me double check and revert. Thank you!

**Comment by hydai** [comment]
> I remember from my morning research that the C dynamic library from WasmEdge release page was ~50MB, which is quite heavy compared to libwasmtime's 17 - are you aware of any thinner versions of it?

Hi,
The official release contains the ahead-of-time compilation (with LLVM inside). So it may take more space. However, if you are looking for a tiny version, we have `wasmedge/slim-runtime` [1], which is the runtime only without the compiler inside.

```
$ file libwasmedge.so.0.0.0
libwasmedge.so.0.0.0: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, BuildID[sha1]=24d0e767d4f29f65dedc1c334e72e2320b6d391c, stripped
$ size libwasmedge.so.0.0.0
   text	   data	    bss	    dec	    hex	filename
1806540	  45848	    904	1853292	 1c476c	libwasmedge.so.0.0.0
~= 1.8M
```

[1]: https://hub.docker.com/repository/docker/wasmedge/slim-runtime

**Comment by juntao** [comment]
Thank you @hydai 

@psarna I think the 1.8MB WasmEdge runtime library is sufficient for your use case. Developers can compile their functions to regular Wasm in any tool they choose. They can further use the `wasmedgec` tool to do AOT compiling before submitting the Wasm file to the database engine. The 1.8MB runtime library can handle both cases. 

Ref: https://wasmedge.org/book/en/quick_start/run_in_aot_mode.html

Also, we do not yet support the proposed "standard" C API. But it could be supported if there is user demand. :)

**Comment by psarna** [comment]
Splendid, thanks guys! 1.8MiB sounds way more aligned with edge use cases indeed, will give it a try

**Comment by psarna** [comment]
v2:
 - .wat source code is now precompiled to a wasm module during initialization, once
 - each valid wasm function is now dynamically registered as a user-defined function, so it's not recompiled on consecutive exeuctions

Still to do: automated tests

**Comment by psarna** [comment]
v3:
 - multiple fixes
 - added CREATE FUNCTION statement
 - added DROP FUNCTION statement

TODO:
 - [ ] automated tests
 - [ ] docs

**Comment by Nelson-He** [comment]
Great works. 
We are working on using  wasm  to supply UDF functions in openGauss database too. 
In our work, we supply an init function to load the wasm mode from local .wasm or .wat file, and parse the file to get the exported function informations. Then users can get the exported function signature intuitively through the tables we supplied.  

We also made a demo to run wasm code in openGuass database, supplied with a docker image to experience.
You can find the project and docker image info below.
[https://github.com/Nelson-He/openGauss-wasm](https://github.com/Nelson-He/openGauss-wasm)

Hope we will keep in touch and exchange the thoughts further more.

**Comment by psarna** [comment]
@penberg please don't merge just yet as it needs more quality Rust tests, covering all types and so on, but I'm marking this PR as ready for review, because the main code is.

**Comment by psarna** [comment]
@penberg I added a few more tests, I'm yet to produce docs (mentioning that the thing is experimental), but code review is welcome!

**Comment by psarna** [comment]
self note: there's a spontaneous double free caused by dropping a wasm function, will investigate tomorrow

**Comment by psarna** [comment]
I added an initial doc explaining Wasm functions and configuration

**Comment by psarna** [comment]
todo based on manual tests: CREATE FUNCTION should accept not only strings, just expressions - that would allow using `readfile` for uploading large .wat files

**Comment by psarna** [comment]
Another todo: we should seriously consider switching to storing raw .wasm binary blobs exclusively, or at least add such support next to .wat files. Wat files are human readable and more verbose, but also more bloated, and require compilation. .wasm on the other hand would allow us to use considerably lighter runtimes (e.g. WasmEdge in slim mode) and would not require a compilation step

**Comment by psarna** [comment]
v4:
 - blobs are supported

**Comment by psarna** [comment]
@penberg @glommer I would appreciate external input on a decision:

Right now, this PR abuses the modularity of libSQL a little, because functions are registered at parsing stage, not later, when the virtual machine executes the statement. Ideally, the only effect of the parsing stage should be vdbe opcodes, to be executed by vdbe later. Implementing it the proper way would take more time and be more intrusive (a new opcode, incompatible with sqlite, would be introduced), but, well, it will be more proper. Do you think I should go ahead and add it to this PR, or move it to a follow-up? Backward compatibility will be preserved for users anyway, but if we prefer this PR to be self-sufficient and solid, it's better if I spend more time on it and add a new opcode on top.

**Comment by psarna** [comment]
v5:
 - opcodes for creating and dropping functions are added
 - EXPLAIN now works, verified with a Rust test

**Comment by psarna** [comment]
@penberg *gentle review ping*

**Comment by psarna** [comment]
While this PR is in review, I'll take a shot at trying to move the Wasm parts of the implementation straight to Rust, thus dropping the libwasmtime dependency. It would be much better developer experience I believe, with no manual steps for downloading dependencies

**Comment by psarna** [comment]
I also plan to make the Rust layer as runtime-agnostic as possible, so hopefully we could easily provide WasmEdge and Wasmer integration later as well

**Comment by penberg** [comment]
@psarna Sorry for asking this so late in the development cycle, but could we refactor this so that the core code only has hooks for _a_ Wasm runtime and move the wasmtime stuff in `ext/wasm`, for example? The reasoning is that the application that embeds libSQL could already have a Wasm runtime (for example, the ChiselStrike runtime already has V8), and it would be great to be able to hook into that. What do you think of this type of design approach?

**Comment by psarna** [comment]
@penberg actually, that's kind of what I'm doing right now - moving the actual implementation of all the bindings to Rust - and then everyone can add their own bindings. `ext/wasm` is already taken for the other way round, compiling libsql to WebAssembly :innocent: but I'll pick another directory name

**Comment by psarna** [comment]
@penberg I did the following v4 change: instead of using wasmtime interface directly, as of now there exists a vendor-agnostic `ext/udf/wasm_bindings.h` header, and a `ext/udf/wasmtime_bindings.c` implementation. The vendor-agnostic header is based on wasmtime's C interface on a 1:1 basis for now, as it's the only working backend for now anyway, and it can be subject to change later. With that, integration with V8 can be done as follows:
1. Produce a `ext/udf/v8_bindings.c` file implementing the interface defined in `ext/udf/wasm_bindings.h`
2. Add it to the build process by editing thousands of Makefile.in, configure.ac and other autoconf files

Here's the commit that introduces the change:
https://github.com/libsql/libsql/pull/45/commits/f7890509006497d610a771c6f2ab9d437741dc89

Opinions?

**Comment by psarna** [comment]
Oh and my idea for later is to build a Rust-based implementation of the `ext/udf/wasm_bindings.h` header, and then we can drop the libwasmtime dependency, because we'll be building a version of it ourselves. That would also be a great opportunity to validate if the interface from `ext/udf/wasm_bindings.h` is convenient enough

**Comment by psarna** [comment]
@penberg v5: there, that's more like it, the implementation now simply needs 2 functions: https://github.com/libsql/libsql/pull/45/commits/aed9bd8f36ce335507785529df68917c7dd5015b , that should be easy enough to port to v8 and native Rust implementation of the Wasmtime backend.

**Comment by psarna** [comment]
note: I'll slightly modify the approach in v6, making wasmtime_bindings.c not rely on any libsql types

**Comment by psarna** [comment]
half-done, but `libsql_run_wasm` is going to need a refactor too - can be done in a follow-up though

**Comment by psarna** [comment]
Now, in order to be able to use WasmEdge as well, the only thing left to do is to port https://github.com/libsql/libsql/blob/f786b293c268e977f579b295d77b9cd107e455a6/src/rust/wasmtime-bindings/src/lib.rs from Wasmtime to WasmEdge - which should be simple enough.

**Comment by psarna** [comment]
@juntao @hydai On top of the existing implementation, I started exploring WasmEdge, it integrates very well! For starters I only implemented ints and doubles as they're trivially mapped to/from libSQL types: 
https://github.com/psarna/libsql/blob/54c495f8c7c8d15edeffd7a2aa6f02d2a7bb80f2/ext/udf/wasmedge_bindings.c
  source: https://github.com/psarna/libsql/commit/54c495f8c7c8d15edeffd7a2aa6f02d2a7bb80f2, https://github.com/psarna/libsql/tree/wasmedge_poc

And it seems to work just fine!
```sql
sqlite> create function fib language wasm as X'0061736d0100000001060160017e017e0302010005030100100619037f01418080c0000b7f00418080c0000b7f00418080c0000b071002066d656d6f727902000366696200000a38013601017e42002101024020004202540d004200210103402000427f7c100020017c21012000427e7c22004201560d000b0b200020017c0b';
Compiled ok!
sqlite> .mode column
sqlite> select fib(10) + fib(11), fib(12);
fib(10) + fib(11)  fib(12)
-----------------  -------
144                144
```

```sh
[sarna@sarna-pc libsql]$ ldd sqlite3 | grep wasmedge
	libwasmedge.so.0 => /usr/local/lib/libwasmedge.so.0 (0x00007f1246c00000)
```

@hydai do you have instructions for compiling slimmed down WasmEdge lib mentioned [here](https://github.com/libsql/libsql/pull/45#issuecomment-1278633479) from source? I'd love to evaluate it. Stripped .so library from default built parameters is 4.3MiB, which is not bad, but for our case, the slimmer the better.

**Comment by hydai** [comment]
Hi @psarna 
Currently, we don't have a document for it. But you can still refer to this workflow file: https://github.com/WasmEdge/WasmEdge/blob/master/.github/workflows/release.yml#L303-L382

Building WasmEdge with `-DWASMEDGE_BUILD_AOT_RUNTIME=OFF` will disable the aot compiler, so it would not include the LLVM framework in the libraries. However, it can still execute the aot compiled wasm/binary because we use `libdl` to handle these files. Also, you can try to add more compiler options to reduce the code size.

Please feel free to let me know if you encounter any other issues, thanks!

**Comment by psarna** [comment]
Great, thanks! `-DWASMEDGE_BUILD_AOT_RUNTIME=OFF` is indeed enough, 1.8MiB with debug symbols and 1.3MiB stripped.

**Comment by psarna** [comment]
@penberg review ping :innocent:

File: src/callback.c [hunk 1]
```diff
@@ -371,8 +371,6 @@ void sqlite3InsertBuiltinFuncs(
     }
   }
 }
-  
-
```
Comment (penberg) [inline src/callback.c:375]: Whitespace noise.

File: src/callback.c [hunk 2]
```diff
@@ -462,7 +460,7 @@ FuncDef *sqlite3FindFunction(
     memcpy((char*)&pBest[1], zName, nName+1);
     for(z=(u8*)pBest->zName; *z; z++) *z = sqlite3UpperToLower[*z];
     pOther = (FuncDef*)sqlite3HashInsert(&db->aFunc, pBest->zName, pBest);
-    if( pOther==pBest ){
+    if(pOther == pBest) {
```
Comment (penberg) [inline src/callback.c:463]: Whitespace noise.

File: src/callback.c [hunk 3]
```diff
@@ -474,6 +472,7 @@ FuncDef *sqlite3FindFunction(
   if( pBest && (pBest->xSFunc || createFlag) ){
     return pBest;
   }
+
```
Comment (penberg) [inline src/callback.c:475]: Whitespace noise.

**Review by penberg — COMMENTED** [review]
I think we should fix this whitespace noise because they will end up causing merge conflicts down the road.

**Comment by psarna** [comment]
@penberg whitespace unnoised!

**Comment by marcobambini** [comment]
Is there any plan to support SQLite aggregate or window functions?
If yes can you please share more details?

**Comment by psarna** [comment]
It's all already possible via the C API (https://www.sqlite.org/c3ref/create_function.html), we don't have any support for the SQL syntax (e.g. `CREATE AGGREGATE` or `CREATE WINDOW FUNCTION`. They would be very nice to have, but also not on our immediate roadmap

**Pull request merged**